packages/api/src/folder/service.ts

```diff
@@ -0,0 +1,114 @@
+import createError from "http-errors";
+import { logger } from "@stela/logger";
+import { db } from "../database";
+import type {
+  Folder,
+  FolderColumnsForUpdate,
+  PatchFolderRequest,
+} from "./models";
+import { requestFieldsToDatabaseFields } from "./helper";
+import { permission } from "../access/permission";
+import { folderAccess } from "./access";
+import { accountService } from "../account/service";
+
+export const getFolderById = async (requestQuery: {
+  folderId: string;
+}): Promise<Folder> => {
+  const folder = await db
+    .sql<Folder>("folder.queries.get_folder_by_id", {
+      folderId: requestQuery.folderId,
+    })
+    .catch((err) => {
+      logger.error(err);
+      throw new createError.InternalServerError("Failed to retrieve folder");
+    });
+
+  if (folder.rows[0] === undefined) {
+    throw new createError.NotFound("Folder not found.");
+  }
+
+  return folder.rows[0];
+};
+
+function buildPatchQuery(patchFolderRequest: FolderColumnsForUpdate): string {
+  const updates = Object.entries(patchFolderRequest)
+    .filter(([key, value]) => value !== undefined && key !== "folderId")
+    .map(([key, _]) => `${key} = :${key}`);
+
+  if (updates.length === 0) {
+    throw new createError.BadRequest("Request cannot be empty");
+  }
+
+  const query = `
+    UPDATE folder
+    SET ${updates.join(", ")}
+    WHERE folderid = :folderId
+    RETURNING folder.folderid AS "folderId"
+  `;
+
+  return query.trim();
+}
+
+const validateCanPatchFolder = async (
+  folderId: string,
+  emailFromAuthToken: string
+): Promise<void> => {
+  const folder = await getFolderById({ folderId });
+  let canEdit = false;
+  const accountArchive = await accountService.getAccountArchive(
+    folder.archiveId,
+    emailFromAuthToken
+  );
+  if (accountArchive && permission.checkCanEditAccountArchive(accountArchive)) {
+    canEdit = true;
+  }
+  if (!canEdit) {
+    const archiveMemberships =
+      await accountService.getCurrentAccountArchiveMemberships(
+        emailFromAuthToken
+      );
+    const editMemberships = await Promise.all(
+      archiveMemberships.map(async (archiveMembership) => {
+        if (permission.checkCanEditAccountArchive(archiveMembership)) {
+          const accessList = await folderAccess.getAccessByFolder(
+            folderId,
+            archiveMembership.archiveId
+          );
+          if (permission.checkCanEdit(accessList)) {
+            return archiveMembership;
+          }
+        }
+        return null;
+      })
+    );
+    if (editMemberships.find((membership) => membership !== null)) {
+      canEdit = true;
+    }
+  }
```

Comment:
I realized in testing that there should be two possible error messages here: the existing 403 if the user is a Viewer or Contributor to the archive where the folder lives, or if they have access to a Viewer or Contributor share of the file, and a 404 in any other case where they can't access it (that is to say, if the user isn't allowed to see the file, we shouldn't confirm for them that it exists).

Reply:
I see, updated the code.

Reply:
I think this covers every case but one - if the caller has view permissions on an archive that has an editable share of the folder, they should get a 403 instead of a 404.